invoice/src/main/resources/org/killbill/billing/invoice/dao/InvoiceSqlDao.sql.stg

```diff
@@ -95,61 +89,78 @@ getInvoiceByAccountRecordIdAfter() ::= <<
    <defaultOrderBy("")>
 >>
 
-searchByAccountOrInvoiceId(ordering) ::= <<
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (<idField("t.")> = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-UNION
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (t.account_id = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-order by <recordIdField("")> <ordering>
-limit :rowCount offset :offset
-;
->>
-
-getSearchByAccountOrInvoiceIdCount() ::= <<
-select count(*) from
-((select
-<allTableFields("t.")>
-from <tableName()> t
-where (<idField("t.")> = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)
-UNION
-(select
-<allTableFields("t.")>
-from <tableName()> t
-where (t.account_id = :searchKey)
-<andCheckSoftDeletionWithComma("t.")>
-<AND_CHECK_TENANT("t.")>)) as count
-;
+invoiceBalanceQuery() ::= <<
+SELECT
+      x.invoice_id
+    , x.tenant_record_id
+    , SUM(x.amount) as balance
+FROM (
+    SELECT
+          i.id as invoice_id
+        , i.tenant_record_id
+        , 0 as amount
+    FROM invoices i
+    WHERE
+        <CHECK_TENANT("i.")>
+
+    UNION ALL
+
+    SELECT
+          ii.invoice_id
+        , ii.tenant_record_id
+        , SUM(ii.amount) as amount
+    FROM invoice_items ii
+    WHERE
+        ii.type in ('TAX', 'EXTERNAL_CHARGE', 'FIXED', 'USAGE', 'RECURRING', 'ITEM_ADJ', 'REPAIR_ADJ', 'PARENT_SUMMARY', 'CBA_ADJ', 'CREDIT_ADJ')
+        <AND_CHECK_TENANT("ii.")>
+    GROUP BY
+          ii.invoice_id
+        , ii.tenant_record_id
+
+    UNION ALL
```

Comment:
Interesting way to write this query...

Reply:
Any concern? It's essentially charges - payments/refunds.  The first subquery which grabs all invoices is to support shell invoices (no invoice items -- at least required for tests).

Reply:
No concerns, LGTM.